### PR TITLE
[CI][Lint] Disable no-else-return check in pylint

### PR DESCRIPTION
**Pull request by Lunderberg**
The `no-else-return` check in pylint has some cases where (to me at least) it leads to less readable code.  In the hypothetical example below, I'd prefer to use if/elif/else to indicate that these are three mutually distinct options, each of which results in similar control flow.  However, the no-else-return check requires writing these as separate if statements.  This reads as though each check is independent, inviting the reader to view the if statements in isolation rather than as a joint structure that exhaustively handles all options.

```python
# Preferred style
def sign(x):
    if x > 0:
        return '+'
    elif x < 0:
        return '-'
    else:
        return ''

# Required by no-else-return check
def sign(x):
    if x > 0:
        return '+'
    if x < 0:
        return '-'
    return ''
```

This is not to say that the if/elif/else style should be used exclusively, but that the style requires by `no-else-return` is not always preferred, and shouldn't be mandated.

cc @Mousius @areusch @driazati

### CODE REVIEW

**Comment by areusch** [comment]
i think for functions which are mapping-style (e.g. given args, return one of a few different expressions) rather than procedural-style (e.g. given args, perform one of a few algorithms), the else-return style can be more readable. because this is strictly making the linter more permissive i'm supportive of it. i'll let others weigh in @apache/tvm-committers

**Review by driazati — APPROVED** [review]

**Comment by zxybazh** [comment]
I’m supportive for this change. I’ve been blocked by this linting check multiple times and it actually makes code less readable. Happy to see what others think!

**Comment by kparzysz-quic** [comment]
I agree with @Lunderberg's sentiment---I think that strict enforcement of style has diminishing returns with the possibility of obscuring author's intent.

File: tests/lint/pylintrc
Comment (AndrewZhaoLuo) [inline tests/lint/pylintrc:85]: Can we split this up into multiple lines to make it more readable (and alphabetize).

e.g.
```
disable=
    too-many-instance-attributes,
    len-as-condition,
    too-few-public-methods,
    anomalous-backslash-in-string,
    no-else-return,
    simplifiable-if-statement,
    too-many-arguments,
    duplicate-code,
    no-name-in-module,
    no-member,
    print-statement,
    parameter-unpackin
```

**Review by AndrewZhaoLuo — APPROVED** [review]
Seems like consensus is here and this is not a super controversial change.

**Comment by Lunderberg** [comment]
> My opinion is that I find the multiple if's far easier to read, as you can discard the context as you go down until you find the eventual output you're looking for

Good point.  I think the distinction in my mind is whether or not the context should be discarded.  In some cases, such as handling a simplified case and early returning, that simplified case no longer needs to be considered.  In other cases, such as handling all options for a variable, that context is helpful to recognize it as a single larger construct.

> So I don't mind either, but given this project has a diverse audience, I'd rather we pick one style and use it consistently so our code ends up looking similar through-out the codebase.

I agree, so long as the consistent style doesn't cause incorrect assumptions for a reader.  For me, seeing a series of `N` if-statements says that there are `2^N` possible paths for control flow, and that they should each be considered, where a if/elif chain with `N` blocks says that there are `N` possible paths for control flow.  While the `return` statement inside each if-statement prevents there from being `2^N` control flow paths, I'd like to avoid that incorrect first impression.

**Comment by junrushao1994** [comment]
Is it good to merge?

**Comment by Lunderberg** [comment]
> Is it good to merge?

I'd like to wait another day or two to make sure the reply addresses @Mousius 's concerns.  Other than that, yes.

**Comment by Mousius** [comment]
> For me, seeing a series of `N` if-statements says that there are `2^N` possible paths for control flow, and that they should each be considered, where a if/elif chain with `N` blocks says that there are `N` possible paths for control flow.

This is a good example of where the ask here is personalised to your way of writing code, gradually over time the end result of this would be removal of all linter rules to allow everyone to write code in their preferred way.

> I'd like to wait another day or two to make sure the reply addresses @Mousius 's concerns. Other than that, yes.

I was voicing my concerns as I think slowly making our code less consistent is a bad idea. In line with the [Code Review Guidelines](https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly), I'm not explicitly requesting a change here, merely leaving a comment for consideration of the committer who takes responsibility in merging this 😸

**Comment by junrushao1994** [comment]
Well...I'm kinda convinced by @Mousius that the change itself is a bit personalized...

**Review by areusch — APPROVED** [review]
my take on project policy is that if we're spending too much time arguing over a topic during code-review and we get inconsistency in the code that's causing problems, we should generalize the topic and have the battle once, then codify the review policy. 

i've elaborated i think clearly enough two classes of functions (procedural and "functional") which are more clearly expressed using different forms of `if-else` and therefore i think this issue would benefit from author/reviewer judgement. i'm okay with approving this as i don't think it's particularly controversial at face value. i'd be happy to reconsider if folks want to raise this again in discussion!

**Comment by areusch** [comment]
@junrushao1994 @Mousius any more comments here? i don't want to just approve and merge without giving ample opportunity for dissent

**Comment by Mousius** [comment]
> i've elaborated i think clearly enough two classes of functions (procedural and "functional") which are more clearly expressed using different forms of if-else and therefore i think this issue would benefit from author/reviewer judgement.

@areusch, I'm still of the opinion that reducing the overall consistency of our codebase based on individuals interpretations of a function style is a bad idea and that need for elaboration will occur in code review repeatedly. The benefit of having a linter is to have a consistent coding style, and not having to discuss it in code reviews. Introducing author/reviewer judgement instead of codified rules means there's the addition of stylistic choices added to the code review process, which could easily have been avoided, and means we increase the diversity of the coding style across the project.

Discussions surrounding linter rules are nearly always about individual coding styles, otherwise we would use the tooling default rules and accept that our individual preferences aren't taken into account (which would also increase consistency across a wider landscape of projects using the same tooling). If we, as a project, want to carry the code review burden of explaining how to format code to contributors then I don't see a reason not to merge this, though the responsible party will likely need to add examples in the code review guidelines to explain when is appropriate to use either case for future contributors given the tooling will no longer provide assistance?

**Comment by Lunderberg** [comment]
Thank you for the feedback on it @Mousius .  I definitely see your point on having a consistent style, and that slow relaxation of linter rules converges on having no linter rules.  I'd also agree about linter rules beings about individual coding style, but would include the tooling defaults in that category as set by the individual coding style of the linter's authors, and so there's a benefit in having project-specific settings.

I think I'd still make a distinction between consistency that adds clarity and consistency that removes clarity.  As a deliberately extreme example, a linter rule requiring all local variables to be single character in alphabetical order of first appearance would make the code much more consistent, at the expense of readability.  While `no-else-return` is nowhere near as extreme as the hypothetical check, I initially proposed disabling the check because it felt to me that it was making a similar tradeoff, without enough consistency to warrant the decrease in readability.

Overall, my preference would be to remove the check, but that the time spent on the discussion has outweighed the benefit from the change, so I'm okay with either outcome as a way of concluding the discussion.

**Comment by areusch** [comment]
we could discuss this a bit further at community meeting, if we'd like to have a time-boxed high-bandwidth discussion. thoughts?

**Comment by Mousius** [comment]
> Thank you for the feedback on it @Mousius . I definitely see your point on having a consistent style, and that slow relaxation of linter rules converges on having no linter rules. I'd also agree about linter rules beings about individual coding style, but would include the tooling defaults in that category as set by the individual coding style of the linter's authors, and so there's a benefit in having project-specific settings.

Just to extend upon your last statement @Lunderberg, I believe some of it is down to personal preference on the part of the linter author but that there's also the specialisation of being the author of such a tool to consider.

> we could discuss this a bit further at community meeting, if we'd like to have a time-boxed high-bandwidth discussion. thoughts?

@areusch, I personally echo the sentiment of @Lunderberg in that _"the time spent on the discussion has outweighed the benefit from the change, so I'm okay with either outcome as a way of concluding the discussion"_.

My only ask is that if this gets merged, we clearly document what we're expecting to occur in each case, within the code review guidelines for future contributors.

**Comment by AndrewZhaoLuo** [comment]
I would like to point out, for C++, our linter allows both styles I believe, so I think there is good argument for consistency in style within codebase.

**Comment by areusch** [comment]
@Mousius that's a reasonable request. @Lunderberg would you be willing to update https://tvm.apache.org/docs/contribute/code_guide.html#python-code-styles with the two types of functions here? my personal preference is to merge this PR, and if no one is vehemently against that and we document it, i think it should be ok to try.

**Comment by Lunderberg** [comment]
> @Mousius that's a reasonable request. @Lunderberg would you be willing to update https://tvm.apache.org/docs/contribute/code_guide.html#python-code-styles with the two types of functions here? my personal preference is to merge this PR, and if no one is vehemently against that and we document it, i think it should be ok to try.

I would, yes, though I wouldn't be able to until after next week.

**Comment by areusch** [comment]
@Lunderberg ok--ping when this gets updated and we can merge.

**Comment by Lunderberg** [comment]
@areusch PR is updated with a section in the code guide to describe the different styles, and a link to this discussion for further information.

**Pull request merged**
